src/components/cc-logs/cc-logs.js

```diff
@@ -0,0 +1,472 @@
+
+  scrollToBottom () {
+    if (this.follow) {
+      this._pin();
+    }
+    else {
+      const element = this._logsRef.value;
+      if (element != null) {
+        element.scrollTop = element.scrollHeight;
+      }
+    }
+  }
+
+  /* endregion */
+
+  firstUpdated (_changedProperties) {
+    if (_changedProperties.has('follow')) {
+      this._pinned = this.follow;
+    }
+  }
+
+  willUpdate (_changedProperties) {
+    if (_changedProperties.has('timestampDisplay')) {
+      this._timestampFormatter = this._resolveTimestampFormatter();
+    }
+    if (_changedProperties.has('follow')) {
+      if (!this.follow) {
+        this._unpin();
+      }
+      else {
+        this._refreshPinnedFromScrollPosition();
+      }
+    }
+  }
+
+  render () {
+    const state = this.logs.state;
+
+    if (state === 'loading') {
+      return html`
+        <div id="loading">
+          <cc-loader></cc-loader>
+        </div>`;
+    }
+
+    if (state === 'error') {
+      return html`
+        <cc-notice intent="danger" message="${i18n('cc-logs.error')}"></cc-notice>
+      `;
+    }
+
+    const __refName = this._logsRef;
+    const logs = this.logs.logs;
+    const lastLogIndex = logs.length - 1;
+    const pin = this._pinned ? {
+      index: logs.length - 1,
+      block: 'end',
+    } : null;
+
+    return html`
+      <div id="logs" tabindex="0" ${ref(__refName)} @scroll=${this._onScroll} @wheel=${this._onMouseWheel}>
+        ${
+          virtualize({
+            scroller: true,
+            items: logs,
+            keyFunction: (it) => it.id ?? it.timestamp,
+            renderItem: (item, index) => {
+              return this._renderLog(
+                item,
+                this.wrapLines,
+                this._timestampFormatter,
+                this.metadataRenderers,
+                this.messageRenderer,
+                index === lastLogIndex,
+              );
+            },
+            layout: {
+              pin,
+            },
+          })
+        }
+      </div>
+    `;
+  }
+
+  /* region Single log line rendering */
+
+  /**
+   * Renders a single line of log.
+   *
+   * @param {Log} log
+   * @param {boolean} wrap
+   * @param {TimestampFormatter} timestampFormatter
+   * @param {Object.<string, MetadataRenderer>} metadataRenderers
+   * @param {MessageRenderer} messageRenderer
+   * @param {boolean} isLast
+   */
+  _renderLog (log, wrap, timestampFormatter, metadataRenderers, messageRenderer, isLast) {
+    return html`
+      <div class="log ${classMap({ wrap: wrap, last: isLast })}">
+        ${this._renderLogTimestamp(log, timestampFormatter)}
+        ${this._renderLogMetadata(log, metadataRenderers)}
+        ${this._renderLogMessage(log, messageRenderer)}
+      </div>
+    `;
+  }
+
+  /**
+   *
+   * @param {Log} log
+   * @param {TimestampFormatter} timestampFormatter
+   */
+  _renderLogTimestamp (log, timestampFormatter) {
+    if (timestampFormatter == null) {
+      return null;
```

Comment:
**nitpick:** should we return `html``;` like we do in `cc-orga-member-list (_renderFormEmailError)` or `cc-email-list (_renderAddressError)`?

```suggestion
      return html``;
```